src/neo4j_genai/retrievers/text2cypher.py

```diff
@@ -0,0 +1,109 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Optional
+
+import neo4j
+from neo4j.exceptions import CypherSyntaxError
+from pydantic import ValidationError
+
+from neo4j_genai.llm import LLM
+from neo4j_genai.prompts import TEXT2CYPHER_PROMPT
+from neo4j_genai.retrievers.base import Retriever
+from neo4j_genai.schema import get_schema
+from neo4j_genai.types import (
+    LLMModel,
+    Neo4jDriverModel,
+    Neo4jSchemaModel,
+    Text2CypherRetrieverModel,
+    Text2CypherSearchModel,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Text2CypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using natural language.
+    Converts a user's natural language query to a Cypher query using an LLM,
+    then retrieves records from a Neo4j database using the generated Cypher query
+    """
+
+    def __init__(
+        self, driver: neo4j.Driver, llm: LLM, neo4j_schema: Optional[str] = None
+    ) -> None:
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            llm_model = LLMModel(llm=llm)
+            neo4j_schema_model = (
+                Neo4jSchemaModel(neo4j_schema=neo4j_schema) if neo4j_schema else None
+            )
+            validated_data = Text2CypherRetrieverModel(
+                driver_model=driver_model,
+                llm_model=llm_model,
+                neo4j_schema_model=neo4j_schema_model,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(validated_data.driver_model.driver)
+        self.llm = validated_data.llm_model.llm
+        self.neo4j_schema = (
+            validated_data.neo4j_schema_model.neo4j_schema
+            if validated_data.neo4j_schema_model
+            else get_schema(validated_data.driver_model.driver)
```

Comment:
I'm thinking if we should try/catch an error here too if get_schema() fails. Perhaps add some clarification of this in the docstring

Reply:
Should handling database errors not be done in the `get_schema` function itself? If `get_schema` fails the only thing we can really do is log then re-raise the error, and it might even just be better to log the error in the `get_schema` function also

Reply:
I think you're right that we should raise exceptions/errors at `get_schema` function level but I think it will also be meaningful here to raise an error here too with a message like `"Failed to get schema for Text2CypherRetriever {error.message}"`. Of course when we refactor to exception handling this will be changed but I think we can raise at both levels

Reply:
Happy to add it in this case but I think we need to come up with a strategy for when to add try-except statements to our code. Adding them at all levels of a call stack just to do logging doesn't seem like best practice.